pkg/controllers/multicluster/globalservice_controller.go

```diff
@@ -0,0 +1,235 @@
+		return fmt.Errorf("unable to list remote endpoint slices for service %v/%v: %v",
+			svc.Name, svc.Namespace, err)
+	}
+
+	repsList := &multiclusterv1.RemoteEndpointSliceList{}
+	if err := r.List(ctx, repsList, client.MatchingLabels{
+		discoveryv1beta1.LabelServiceName: svc.Name,
+	}); err != nil {
+		return fmt.Errorf("unable to list remote endpoint slices for service %v/%v: %v",
+			svc.Name, svc.Namespace, err)
+	}
+
+	repsMap := map[string]string{}
+	for _, reps := range repsList.Items {
+		if reps.DeletionTimestamp.IsZero() {
+			repsMap[reps.Name] = reps.Name
+		}
+	}
+
+	for _, eps := range epsList.Items {
+		if _, exist := repsMap[eps.Name]; !exist {
+			if err := client.IgnoreNotFound(r.Delete(ctx, &eps)); err != nil {
+				return fmt.Errorf("unable to delete endpoint slice %v/%v: %v", eps.Name, eps.Namespace, err)
+			}
+		}
+	}
+
+	for index := range repsList.Items {
+		reps := &repsList.Items[index]
+
+		if reps.Spec.RemoteService.Namespace != svc.Namespace {
+			continue
+		}
+
+		var endpointSlice = &discoveryv1beta1.EndpointSlice{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      reps.Name,
+				Namespace: reps.Spec.RemoteService.Namespace,
+			},
+		}
+
+		_, err := controllerutil.CreateOrPatch(ctx, r, endpointSlice, func() error {
+			if !endpointSlice.DeletionTimestamp.IsZero() {
+				return fmt.Errorf("endpoint slice %s is terminating, can not be updated", endpointSlice.Name)
+			}
+
+			endpointSlice.Labels = reps.Labels
+			endpointSlice.Annotations = reps.Annotations
+			endpointSlice.Endpoints = reps.Spec.Endpoints
+			endpointSlice.AddressType = reps.Spec.AddressType
+			endpointSlice.Ports = reps.Spec.Ports
+
+			if !metav1.IsControlledBy(endpointSlice, svc) {
+				// this is the same as EndpointSlices create by kube-controller-manager
+				if err := controllerutil.SetControllerReference(svc, endpointSlice, r.Scheme()); err != nil {
+					return wrapError("unable to set owner reference", err)
+				}
+			}
+
+			return nil
+		})
+		if err != nil {
+			return fmt.Errorf("unable to update endpoint slice %v: %v", endpointSlice.Name, err)
+		}
+	}
+
+	return nil
+}
+
+func (r *GlobalServiceReconciler) SetupWithManager(mgr ctrl.Manager) (err error) {
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(ControllerGlobalService).
+		For(&corev1.Service{},
+			builder.WithPredicates(
+				globalServicePredicate(),
+				predicate.Or(
+					&predicate.GenerationChangedPredicate{},
+					&predicate.AnnotationChangedPredicate{},
+				),
+			),
+		).
+		Watches(&source.Kind{Type: &multiclusterv1.RemoteEndpointSlice{}}, handler.EnqueueRequestsFromMapFunc(
+			func(object client.Object) []reconcile.Request {
+				reps, ok := object.(*multiclusterv1.RemoteEndpointSlice)
+				if !ok {
+					return nil
+				}
+
+				return []reconcile.Request{
+					{
+						NamespacedName: types.NamespacedName{
+							Name:      reps.Spec.RemoteService.Name,
+							Namespace: reps.Spec.RemoteService.Namespace,
+						},
+					},
+				}
+			})).
```

Comment:
just another question, should *local* endpoint slice be watched here to make sure that endpoint slices binded on remote endpoint slices will be repaired after unexpected update or deletion?

Reply:
done